pkg/controller/master/upgrade/common.go

```diff
@@ -71,173 +108,296 @@ func markComplete(upgrade *harvesterv1.Upgrade) {
+						upgradeJobTypeLabel:            jobType,
+					},
+				},
+				Spec: corev1.PodSpec{
+					HostIPC:       true,
+					HostPID:       true,
+					HostNetwork:   true,
+					DNSPolicy:     corev1.DNSClusterFirstWithHostNet,
+					RestartPolicy: corev1.RestartPolicyNever,
+					Containers: []corev1.Container{
+						{
+							Name:    "apply",
+							Image:   fmt.Sprintf("%s:%s", upgradeImageRepository, imageVersion),
+							Command: []string{"upgrade_node.sh"},
+							Args:    []string{jobType},
+							Env: []corev1.EnvVar{
+								{
+									Name:  "HARVESTER_UPGRADE_NAME",
+									Value: upgrade.Name,
+								},
+								{
+									Name:  "HARVESTER_UPGRADE_NODE_NAME",
+									Value: nodeName,
+								},
+								{
+									Name: "HARVESTER_UPGRADE_POD_NAME",
+									ValueFrom: &corev1.EnvVarSource{
+										FieldRef: &corev1.ObjectFieldSelector{
+											FieldPath: "metadata.name",
+										},
+									},
+								},
+							},
+							VolumeMounts: []corev1.VolumeMount{
+								{Name: "host-root", MountPath: "/host"},
+							},
+							SecurityContext: &corev1.SecurityContext{
+								Privileged: &privileged,
+								Capabilities: &corev1.Capabilities{
+									Add: []corev1.Capability{
+										corev1.Capability("CAP_SYS_BOOT"),
+									},
+								},
+							},
+						},
+					},
+					ServiceAccountName: "harvester",
+					Affinity: &corev1.Affinity{
+						NodeAffinity: &corev1.NodeAffinity{
+							RequiredDuringSchedulingIgnoredDuringExecution: &corev1.NodeSelector{
+								NodeSelectorTerms: []corev1.NodeSelectorTerm{{
+									MatchExpressions: []corev1.NodeSelectorRequirement{{
+										Key:      corev1.LabelHostname,
+										Operator: corev1.NodeSelectorOpIn,
+										Values: []string{
+											nodeName,
+										},
+									}},
+								}},
+							},
+						},
+					},
+					Tolerations: []corev1.Toleration{
+						{
+							Key:      corev1.TaintNodeUnschedulable,
+							Operator: corev1.TolerationOpExists,
+							Effect:   corev1.TaintEffectNoSchedule,
+						},
+						{
+							Key:      "kubevirt.io/drain",
+							Operator: corev1.TolerationOpExists,
+							Effect:   corev1.TaintEffectNoSchedule,
+						},
+						{
+							Key:      labelCriticalAddonsOnly,
+							Operator: corev1.TolerationOpExists,
+						},
+						{
+							Key:      corev1.TaintNodeUnreachable,
+							Operator: corev1.TolerationOpExists,
+							Effect:   corev1.TaintEffectNoExecute,
+						},
+						{
+							Key:      labelArch,
+							Operator: corev1.TolerationOpEqual,
+							Effect:   corev1.TaintEffectNoSchedule,
+							Value:    "amd64",
+						},
+						{
+							Key:      labelArch,
+							Operator: corev1.TolerationOpEqual,
+							Effect:   corev1.TaintEffectNoSchedule,
+							Value:    "arm64",
+						},
+						{
+							Key:      labelArch,
+							Operator: corev1.TolerationOpEqual,
+							Effect:   corev1.TaintEffectNoSchedule,
+							Value:    "arm",
+						},
```

Comment:
Found this part duplicate over and over again. Can we extract it?

Reply:
Thanks. Fixed in 1b5ea72